### DISCUSSION TITLE
log and alert for voltage out of range during outage ?

### DISCUSSION THREAD

**Discussion by nworbekim** [jgyates/genmon#1089] [Questions] [answered]
### Checklist

- [X] I made sure that there are *no existing issues or discussions* - [open issues](https://github.com/jgyates/genmon/issues),[closed issues](https://github.com/jgyates/genmon/issues?q=is%3Aissue+is%3Aclosed),[open discussion](https://github.com/jgyates/genmon/discussions), [closed discussion](https://github.com/jgyates/genmon/discussions/categories/questions?discussions_q=is%3Aclosed) - which may answer my question or that I could contribute my information to.
- [X] I have read the [Known Issues Wiki Page](https://github.com/jgyates/genmon/wiki/Appendix-D-Known-Issues) and my question / issue is not listed.
- [X] I have provided the version of the software and the operating system in my post.
- [X] If I have an issue with genmon not communicating with the controller, I have read and followed the [serial troubleshooting guide](https://github.com/jgyates/genmon/wiki/3.6---Serial-Troubleshooting).

### Version

1.19.02

### Have you submitted your log files from the About page? Note that outbound email must be working for logs to be submitted.

I have submitted logs

### What is/are your question(s)?

Genmon 1.19.02
Raspberry Pi 3 B
Debian 12.2 (from /etc/debian_conf)

Apologies in advance if I missed this issue in the discussion archives.

My Generac (Nexus) generator recently has had trouble maintaining constant voltage.   The voltage swings high then low (300 V to 195 V).   I found this by looking at Genmon's dashboard when the lights were flickering, dimming and going very bright during a recent outage.   The maintenance tech didn't find anything wrong when they came to check the next day.   I'd like to log the voltage data in case this happens again.  This is for Genmon 1.19.02 on Debian 12.2

Two questions:
1 - what do you think of adding this feature to Genmon?  The concept is sending additional notices of voltage out of range during an outage.
2 - for a DIY monitoring solution, which http status requests would you recommend I use to log and alert on generator voltage?

**Comment by jgyates** [comment, marked as answer]
That sounds good in principal, however how is this different from the alarm conditions built into the controller for overspeed, underspeed, under voltage and over voltage? These should cover most situations. Your logs show that you did get alarms for overspeed and undervoltage recently. Overspeed alarms also can indicate over voltage. Did you run into a situation 

If you want to log the voltage data you can use this URL (adjust ip and https for your settings):

     http://192.168.1.10:8000/cmd/status_json

The attached file is a program that you could use to log the generator output voltage to a file. It uses the above URL to output the time date, output voltage and engine status. This could be logged to a file or modified to only write the data if the engine is running, or your could change the frequency the data is polled. 

Since you can not attached a python file you will need to rename this to voltage.py. 

[voltage.py.txt](https://github.com/jgyates/genmon/files/14706720/voltage.py.txt)

This thread has a similar approach:  https://github.com/jgyates/genmon/issues/1067

**Reply by nworbekim** [reply]
Thank you for the code.  It is very helpful.  I made some changes to check  the return values  and vary the sleep time so it checks every second when the engine is running.

Where do you see recent entries for over speed, under voltage and/or over voltage?   The Feb 28, 2024 outage is when I observed the voltage problems.  In Genmon's UI, the most recent log entry is for overspeed shows June 2022, under voltage was last recorded in January 2022.   In 2022 I had the throttle servo control replaced and that fixed the overspeed and under voltage errors.

For the Feb 28 2024 log entries (over speed, over voltage, under voltage).  
- The engine's RPMs were stable at 3600 rpm when the voltage was fluctuating.  
- I will check the system's on-board logs when I have access.  I don't recall seeing any under or over-voltage log entries.   Genmon didn't record any log entries for under or over voltage.    
- During the voltage changes, I was watching the dashboard voltage dial and it matched the lights dimming and brightening.

Here's the modified copy of your python I'm using to log the voltage:
[voltage-2.py.txt](https://github.com/jgyates/genmon/files/14718640/voltage-2.py.txt)

**Comment by nworbekim** [comment]
Hi,

I wanted to give you an update on the under/over voltage issues I'm having.   

The generator does not log under or over voltage alarms but the voltage data shows momentary spikes and drops far exceeding the normal range.   Attached is a picture of one of graphs showing the voltage dropping to 190 and spiking to 300 volts.  The lights in the house dim and brighten in sync with the voltage changes.

Our generator maintenance company was able to observe the problem during our last outage and believes we have failing voltage regulator which they plan to replace soon. 

I doubt this is a common problem but it might be useful for others if there's some sort of monitor or wiki page entry to tell others that momentary voltage changes may not trigger the generator's alarms.

Regards
Mike

<img width="971" alt="Screenshot 2024-04-11 at 08 59 59" src="https://github.com/jgyates/genmon/assets/5480401/3828af7d-ed1d-4271-ac25-72432a9a3ae9">
[woltage-2024-04-05b.csv](https://github.com/jgyates/genmon/files/14946435/woltage-2024-04-05b.csv)

**Reply by jgyates** [reply]
Thanks for the update. I added this monitoring request to the feature list: https://github.com/jgyates/genmon/blob/master/feature_requests.md

I wonder if not alerting to over voltage conditions is unique to the Nexus controller, or if you have a unique failure that for some reason is not being alerted. If it is the Nexus controller and not a failure, I would be surprised if that is the case for Evolution too as this is a pretty basic check.

**Reply by nworbekim** [reply]
Thanks.  Your program and GitHub site have been incredibly useful;  Your work has given us insight into our generator's condition and  helped us with utility power problems (we were getting regular 7 second outages -- when we reported them the electric company's engineer immediately recognized they had a problem and fixed it).

For the under/over voltage alarms, my speculation is the generator alarms are not triggered by momentary dips or spikes.   A couple years ago, our generator logged under voltage alarms--so it was working ...   If the Nexus alarm trigger is in the voltage regulator then that could also explain why I'm not seeing voltage alarms now.

**Comment by nworbekim** [comment]
FYI - finally got the controller board replaced (brushes serviced, etc) and now the generator's voltage is staying within a few volts of 240V under load.  I'm still using the logging scripts until I have more confidence in the repairs

**Reply by jgyates** [reply]
Thanks for the update
